homeassistant/components/fritz/common.py

```diff
@@ -0,0 +1,266 @@
+        return f"{DOMAIN}-device-update"
+
+    def _update_info(self):
+        """Retrieve latest information from the FRITZ!Box."""
+        if not self.success:
+            return None
+
+        return self.fritzhosts.get_hosts_info()
+
+    def scan_devices(self, now: Optional[datetime] = None) -> None:
+        """Scan for new devices and return a list of found device ids."""
+
+        _LOGGER.debug("Checking devices for Fritz!Box router %s", self.host)
+
+        new_device = False
+        for known_host in self._update_info():
+            if not known_host.get("mac"):
+                continue
+
+            dev_mac = known_host["mac"]
+            dev_name = known_host["name"]
+            dev_ip = known_host["ip"]
+            dev_home = known_host["status"]
+
+            dev_info = Device(dev_mac, dev_ip, dev_name)
+
+            if dev_mac in self._devices:
+                self._devices[dev_mac].update(dev_info, dev_home)
+            else:
+                device = FritzScannerEntity(dev_mac)
+                device.update(dev_info, dev_home)
+                self._devices[dev_mac] = device
+                new_device = True
+
+        async_dispatcher_send(self.hass, self.signal_device_update)
+        if new_device:
+            async_dispatcher_send(self.hass, self.signal_device_new)
+
+    def _fetch_device_info(self):
+        """Fetch device info."""
+        info = self.connection.call_action("DeviceInfo:1", "GetInfo")
+        return {
+            "identifiers": {
+                # Serial numbers are unique identifiers within a specific domain
+                (DOMAIN, self.unique_id)
+            },
+            "name": info.get("NewModelName"),
+            "manufacturer": "AVM",
+            "model": info.get("NewModelName"),
+            "sw_version": info.get("NewSoftwareVersion"),
+        }
+
+
+class FritzScannerEntity:
+    """FritzScanner device."""
+
+    def __init__(self, mac, name=None):
+        """Initialize device info."""
+        self._mac = mac
+        self._name = name
+        self._ip_address = None
+        self._last_activity = None
+        self._connected = False
+        self._icon = None
+
+    def update(self, dev_info, dev_home):
+        """Update device info."""
+        utc_point_in_time = dt_util.utcnow()
+        if not self._name:
+            self._name = dev_info.name or self._mac.replace(":", "_")
+        self._connected = dev_home
+
+        if not self._connected:
+            self._ip_address = None
+            self._icon = "mdi:lan-disconnect"
+        else:
+            self._last_activity = utc_point_in_time
+            self._ip_address = dev_info.ip
+            self._icon = "mdi:lan-connect"
+
+    @property
+    def is_connected(self):
+        """Return connected status."""
+        return self._connected
+
+    @property
+    def mac(self):
+        """Get MAC address."""
+        return self._mac
+
```

Comment:
If we can get the `hostname` this can be used to accelerate dhcp discovery when the integration is setup https://developers.home-assistant.io/docs/core/entity/device-tracker/#dhcp-discovery

Reply:
Renamed mac to `mac_address` and name to `hostname`